modin/engines/base/io/file_reader.py

```diff
@@ -116,6 +116,17 @@ def file_size(cls, f):
         f.seek(cur_pos, os.SEEK_SET)
         return size
 
+    @classmethod
+    def lines_amount(cls, f, from_begin=False):
```

Comment:
How does this work with quotes?

Reply:
nohow, it's just line counting

actually we don't need to count rows of dataframe (with considering quotes), because all that logic encapsulated in `call_deploy`.
Let's say that we have file with 6 lines (do not counting header), 3 rows and we want to read it with 2 partitions, so each partition should read 6/2=3 lines:
```
col1,col2,col3
"very very                      <---- this peace for partition №1
long string
that do not
-----------------------------
fit into partition",1,2     <---- and that peace for partition №2
3,4,5
6,7,8
```
First partition will read 3 lines and read one more line to complete row thanks to this check:
https://github.com/dchigarev/modin/blob/8b5866337ce3e5e14b8c00c39e2ed465b4417a0e/modin/engines/base/io/text/text_file_reader.py#L30-L39

So, second partition still want to read 3 lines, but there is only 2 left (because first partition readed 1 extra line to complete row), but it's not a problem, because `islice` feels okay about creating sequences which is longer than source object and will just return 2 lines as we wanted

Reply:
found out that it's very useful to know amount of rows that every partition will read, instead of lines (this is critical for example for that issue #1067). 

Also found out that it's really challenging to write fast regexp that would count amount of `\n` that not in quotes, so for now I'm gonna use that approach with line counting in a for loop.

Here are time measurements of three approaches of rows counting, one with a for loop and other two with different regexpes:

<details><summary>reproducer</summary>


```python
import numpy as np
import string
import csv
import os
import re
from timeit import default_timer as timer

RAND_LOW = -100
RAND_HIGH = 100
N = 500
M = 128
TEST_FILENAME = os.path.abspath(f"int_dataset-{N},{M},{RAND_LOW},{RAND_HIGH}.csv")


def generate_data_file(filename, row_n, col_n):
    letters = list(string.ascii_letters)
    columns = [
        "".join(np.random.choice(letters) for _ in np.arange(10))
        for _ in np.arange(col_n)
    ]

    with open(filename, "w") as file:
        writer = csv.writer(file, delimiter=",")
        writer.writerow(columns)
        for _ in np.arange(row_n):
            writer.writerow(np.random.randint(RAND_LOW, RAND_HIGH, col_n))


if __name__ == "__main__":
    if not os.path.exists(TEST_FILENAME):
        generate_data_file(TEST_FILENAME, N, M)
    path = TEST_FILENAME

    # regexp_0: if match has nonempty group №4, then this is
    # a new row. This method has disadvantage that we must iterate
    # over all of the matches and count only these which contains not
    # empty group №4
    regexp_0_groups = b"((\"[^\"]*\")|(.*))(.*$)"

    # regexp_1: maths all end of lines that's outside of quotes.
    # Disadvantage: cause of regex complexity it works really slow
    regexp_1_findall = b"$(?=(?:[^\"]*\"[^\"]*\")*[^\"]*\Z)"

    with open(path, "rb") as f:
        t1 = timer()
        outside_quotes = True
        nrows = 0
        for line in f:
            if line.count(b"\"") % 2:
                outside_quotes = not outside_quotes
            if outside_quotes:
                nrows += 1
        print("rows counting in a loop:", timer() - t1)

    with open(path, "rb") as f:
        t1 = timer()
        content = f.read()
        nrows = 0
        it = re.finditer(regexp_0_groups, content, re.MULTILINE)
        for x in it:
            g1, g2 = x.group(4), x.group(1)
            if (g1 is not None and g1 != b"") or (g2 is not None and g2 != b""):
                nrows += 1
        print("groups regexp: ", timer() - t1)

    with open(path, "rb") as f:
        t1 = timer()
        content = f.read()
        nrows = len(re.findall(regexp_1_findall, content, re.MULTILINE))
        print("findall", timer() - t1)

```


</details>

```
5000 rows:
for loop counting: 0.0044s
regexp_0: 0.0123s
regexp_1: 20.2422s

500,000 rows:
for loop counting: 0.4450s
regexp_0: 1.2126s
regexp_1: >100s

700,000 rows:
for loop counting: 0.6004s
regexp_0: 1.7894s
regexp_1: >100s

5,500,000 rows (~2.5 gb):
for loop counting: 4.9384s
regexp_0: 13.8057s
regexp_1: >100s
```

so in the next commit, I'll try to integrate rows counting via for loop